File: snappy-pi/Snappy.cs
// Copyright (c) Christopher Currens.  Licensed under the Apache 2.0 License (http://www.apache.org/licenses/LICENSE-2.0.html)

using System;

namespace Snappy.Net
{
    public static class Snappy
    {
        public static long GetUncompressedLength(byte[] compressed)
        {
            if (compressed == null)
            {
                throw new ArgumentNullException("compressed");
            }

            ulong result;
            if (!NativeMethods.GetUncompressedLength(compressed, (ulong)compressed.Length, out result))
            {
                throw new InvalidSnappyDataException("Compressed length couldn't be parsed from source data");
            }
            
            if (result > long.MaxValue)
            {
                // This is most likely corrupt data...can't uncompress 8EB in memory in 2013.
                throw new InvalidSnappyDataException("Uncompressed length is greater than Int64.MaxValue");
            }

            return (long)result;
        }

        public static bool IsValidCompressedBuffer(byte[] compressed)
        {
            if (compressed == null)
            {
                throw new ArgumentNullException("compressed");
            }

            var isComp = NativeMethods.IsValidCompressedBuffer(compressed, (ulong)compressed.Length);
            return isComp;
        }

        public static int MaxCompressedLength(int sourceLength)
        {
            if (sourceLength < 0)
            {
                throw new ArgumentOutOfRangeException("sourceLength");
            }

            var result = MaxCompressedLength((long) sourceLength);

            if (result > int.MaxValue)
            {
                throw new SnappyException("Max compressed length is greater than Int32.MaxValue.  Use the long overload instead.");
            }

            return (int) result;
        }

        public static long MaxCompressedLength(long sourceLength)
        {
            if (sourceLength < 0)
            {
                throw new ArgumentOutOfRangeException("sourceLength");
            }

            try
            {
                return checked(32 + sourceLength + (sourceLength / 6));
            }
            catch (OverflowException)
            {
                // Hitting an overflow is likely programmer's error...it would be a massive byte[] allocation.
                throw new SnappyException("Max compressed length is greater than Int64.MaxValue.");
            }
        }

        public static byte[] Uncompress(byte[] compressed)
        {
            if (compressed == null)
            {
                throw new ArgumentNullException("compressed");
            }
            
            var uncompressedLength = GetUncompressedLength(compressed);
            var buf = new byte[uncompressedLength];

            if (!Uncompress(compressed, buf))
            {
                throw new InvalidSnappyDataException("Uncompress failed.  Data is likely corrupt.");
            }

            return buf;
        }

        public static bool Uncompress(byte[] compressed, byte[] uncompressed)
        {
            if (compressed == null)
            {
                throw new ArgumentNullException("compressed");
            }

            if (uncompressed == null)
            {
                throw new ArgumentNullException("uncompressed");
            }

            return NativeMethods.RawUncompress(compressed, (ulong)compressed.Length, uncompressed);
        }

        public static byte[] Compress(byte[] input)
        {
            if (input == null)
            {
                throw new ArgumentNullException("input");
            }

            ulong compLen;

            var maxCompLen = MaxCompressedLength(input.Length);
            var compBuffer = new byte[maxCompLen];

            NativeMethods.RawCompress(input, (ulong)input.Length, compBuffer, out compLen);

            if (compLen == 0)
            {
                throw new SnappyException("Compression failed for unknown reasons.");
            }
            
            // TODO: Do this right, with long instead of int casts.  But who is going to compress > 2GB of data in memory...well at least now?
            var b = new byte[compLen];
            Buffer.BlockCopy(compBuffer, 0, b, 0, (int)compLen);

            return b;
        }
    }
}

File: Snappy.Net-CppCli/snappy.cpp

#include "snappy.h"
#include "exceptions.h"
#include "..\snappy\snappy.h"

namespace Snappy { namespace Net {

    long Snappy::GetUncompressedLength(array<Byte>^ compressed)
    {
        if(compressed == nullptr) {
            throw gcnew ArgumentNullException("compressed");
        }

        pin_ptr<Byte> lpInput = &compressed[0];

        size_t outLength;
        if(!snappy::GetUncompressedLength((char*)lpInput, compressed->Length, &outLength)) {
            throw gcnew InvalidSnappyDataException("Compressed length couldn't be parsed from source data");
        }

        if(outLength > Int64::MaxValue) {
            // Corrupt data...or somehow 8EB was compressed in memory. :x
            throw gcnew InvalidSnappyDataException("Uncompressed length is greater than Int64.MaxValue");
        }

        return (long)outLength;
    }

    bool Snappy::IsValidCompressedBuffer(array<Byte>^ compressed)
    {
        if(compressed == nullptr) {
            throw gcnew ArgumentNullException("compressed");
        }

        pin_ptr<Byte> lpInput = &compressed[0];
        return snappy::IsValidCompressedBuffer((char*)lpInput, compressed->Length);
    }

    long Snappy::MaxCompressedLength(long sourceLength)
    {
        if(sourceLength < 0) {
            throw gcnew ArgumentOutOfRangeException("sourceLength");
        }

        auto result = (long)snappy::MaxCompressedLength(sourceLength);
        if(result < 0) {
            throw gcnew SnappyException("Max compressed length is greater than Int64.MaxValue");
        }

        return result;
    }

    array<Byte>^ Snappy::Uncompress(array<Byte>^ compressed)
    {
        if(compressed == nullptr) {
            throw gcnew ArgumentNullException("compressed");
        }

        // Calling our method instead of native will do range and validation checks for us.
        long uncompressedLength = GetUncompressedLength(compressed);

        auto buffer = gcnew array<Byte>(uncompressedLength);
        
        if(!Uncompress(compressed, buffer))
        {
            throw gcnew InvalidSnappyDataException("Failed to uncompress.  The data is likely invalid or corrupt.");
        }

        return buffer;
    }

    bool Snappy::Uncompress(array<Byte>^ compressed, array<Byte>^ uncompressed)
    {
        if(compressed == nullptr) {
            throw gcnew ArgumentNullException("compressed");
        }

        if(uncompressed == nullptr) {
            throw gcnew ArgumentNullException("uncompressed");
        }

        pin_ptr<Byte> lpInput = &compressed[0];
        pin_ptr<Byte> lpOutput = &uncompressed[0];

        return snappy::RawUncompress((char*)lpInput, compressed->Length, (char*)lpOutput);
    }

    array<Byte>^ Snappy::Compress(array<Byte>^ inputData)
    {
        if(inputData == nullptr) {
            throw gcnew ArgumentNullException("inputData");
        }

        // Create a buffer for the 
        long maxCompressedLen = MaxCompressedLength(inputData->Length);
        auto buffer = gcnew array<Byte>(maxCompressedLen);

        long compressedSize = Compress(inputData, buffer);

        auto resultBuf = gcnew array<Byte>(compressedSize);
        Buffer::BlockCopy(buffer, 0, resultBuf, 0, compressedSize);

        return resultBuf;
    }

    long Snappy::Compress(array<Byte>^ inputData, array<Byte>^ outputBuffer) {
        pin_ptr<Byte> lpInput = &inputData[0];
        pin_ptr<Byte> lpOutput = &outputBuffer[0];

        size_t compressedSize;
        snappy::RawCompress((char*)lpInput, inputData->Length, (char*)lpOutput, &compressedSize);

        // This should never be 0.  It will at least be the size of a header.
        if(compressedSize == 0) {
            throw gcnew SnappyException("Compressed failed for unknown reasons!");
        }

        return compressedSize;
    }
}}